src/bp/core/services/ghost/service.ts

```diff
@@ -55,14 +56,21 @@ export class GhostService {
       throw new Error(`Invalid botId "${botId}"`)
     }
 
-    return new ScopedGhostService(
+    if (this._scopedGhosts.has(botId)) {
+      return this._scopedGhosts.get(botId)!
+    }
+
+    const scopedGhost = new ScopedGhostService(
```

Comment:
the performance impact is mostly on the garbage collector. We're creating one instance per bot, instead of one for each incoming events, which has a big impact under heavy load